File: src/manifest.rs
use std::{collections::BTreeSet, path::Path, process::Command};

use anyhow::{Context, Result, anyhow, bail};
use semver::Version;
use serde_json::Value;

#[derive(Debug, Clone)]
pub(crate) struct ManifestDependency {
    pub(crate) identity: String,
    pub(crate) kind: ManifestDependencyKind,
    pub(crate) location: String,
    pub(crate) requirement: Requirement,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub(crate) enum ManifestDependencyKind {
    SourceControl,
    Registry,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct ManifestFileSystemDependency {
    pub(crate) identity: String,
    pub(crate) name: String,
    pub(crate) path: String,
}

#[derive(Debug, Clone)]
pub(crate) enum Requirement {
    Exact(Version),
    Range { lower: Version, upper: Version },
    Revision(String),
    Branch(String),
}

pub(crate) fn dump_package(package_dir: &Path, disable_sandbox: bool) -> Result<Value> {
    let output = dump_package_json(package_dir, disable_sandbox)?;
    serde_json::from_slice(&output).context("failed to parse dump-package JSON")
}

pub(crate) fn dump_package_json(package_dir: &Path, disable_sandbox: bool) -> Result<Vec<u8>> {
    let mut command = Command::new("swift");
    command.arg("package");
    if disable_sandbox {
        command.arg("--disable-sandbox");
    }
    let output = command
        .arg("dump-package")
        .current_dir(package_dir)
        .output()
        .context("failed to run swift package dump-package")?;
    if !output.status.success() {
        bail!(
            "swift package dump-package failed:\n{}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
    Ok(output.stdout)
}

pub(crate) fn parse_manifest_dependencies(manifest: &Value) -> Result<Vec<ManifestDependency>> {
    let mut dependencies = Vec::new();
    let Some(items) = manifest.get("dependencies").and_then(Value::as_array) else {
        return Ok(dependencies);
    };

    for item in items {
        if let Some(source_control) = item.get("sourceControl").and_then(Value::as_array) {
            for dependency in source_control {
                let identity = dependency
                    .get("identity")
                    .and_then(Value::as_str)
                    .ok_or_else(|| anyhow!("sourceControl dependency is missing identity"))?
                    .to_string();
                let location = parse_source_control_location(dependency)
                    .ok_or_else(|| anyhow!("{identity} is missing source-control location"))?;
                let requirement = parse_requirement(
                    dependency
                        .get("requirement")
                        .ok_or_else(|| anyhow!("{identity} is missing requirement"))?,
                )
                .with_context(|| format!("failed to parse requirement for {identity}"))?;
                dependencies.push(ManifestDependency {
                    identity,
                    kind: ManifestDependencyKind::SourceControl,
                    location,
                    requirement,
                });
            }
        }

        if let Some(registry) = item.get("registry").and_then(Value::as_array) {
            for dependency in registry {
                let identity = dependency
                    .get("identity")
                    .and_then(Value::as_str)
                    .ok_or_else(|| anyhow!("registry dependency is missing identity"))?
                    .to_string();
                let requirement = parse_requirement(
                    dependency
                        .get("requirement")
                        .ok_or_else(|| anyhow!("{identity} is missing requirement"))?,
                )
                .with_context(|| format!("failed to parse requirement for {identity}"))?;
                dependencies.push(ManifestDependency {
                    location: identity.clone(),
                    identity,
                    kind: ManifestDependencyKind::Registry,
                    requirement,
                });
            }
        }
    }

    Ok(dependencies)
}

fn parse_source_control_location(dependency: &Value) -> Option<String> {
    dependency
        .pointer("/location/remote/0/urlString")
        .and_then(Value::as_str)
        .or_else(|| {
            dependency
                .pointer("/location/local/0")
                .and_then(Value::as_str)
        })
        .map(ToString::to_string)
}

pub(crate) fn parse_required_manifest_dependencies(
    manifest: &Value,
) -> Result<Vec<ManifestDependency>> {
    let dependencies = parse_manifest_dependencies(manifest)?;
    let references = active_dependency_references(manifest);
    if references.is_empty() {
        return Ok(Vec::new());
    }

    Ok(dependencies
        .into_iter()
        .filter(|dependency| {
            dependency_reference_names(dependency)
                .into_iter()
                .any(|name| references.contains(&name))
        })
        .collect())
}

fn active_dependency_references(manifest: &Value) -> BTreeSet<String> {
    let mut references = BTreeSet::new();
    let Some(targets) = manifest.get("targets").and_then(Value::as_array) else {
        return references;
    };
    let target_names = targets
        .iter()
        .filter_map(|target| target.get("name").and_then(Value::as_str))
        .collect::<BTreeSet<_>>();
    let mut pending_targets = manifest
        .get("products")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|product| product.get("targets").and_then(Value::as_array))
        .flatten()
        .filter_map(Value::as_str)
        .collect::<Vec<_>>();
    pending_targets.extend(
        targets
            .iter()
            .filter(|target| target.get("type").and_then(Value::as_str) == Some("test"))
            .filter_map(|target| target.get("name").and_then(Value::as_str)),
    );
    let mut visited_targets = BTreeSet::new();

    while let Some(target_name) = pending_targets.pop() {
        if !visited_targets.insert(target_name) {
            continue;
        }
        let Some(target) = targets
            .iter()
            .find(|target| target.get("name").and_then(Value::as_str) == Some(target_name))
        else {
            continue;
        };
        let Some(dependencies) = target.get("dependencies").and_then(Value::as_array) else {
            continue;
        };
        for dependency in dependencies {
            if let Some(product) = dependency.get("product").and_then(Value::as_array) {
                let product_name = product.first().and_then(Value::as_str);
                let package_name = product.get(1).and_then(Value::as_str);
                if let Some(package_name) = package_name.or(product_name) {
                    references.insert(normalize_dependency_reference(package_name));
                }
            }
            if let Some(by_name) = dependency.get("byName").and_then(Value::as_array) {
                if let Some(name) = by_name.first().and_then(Value::as_str) {
                    if target_names.contains(name) {
                        pending_targets.push(name);
                    } else {
                        references.insert(normalize_dependency_reference(name));
                    }
                }
            }
        }
    }

    references
}

fn dependency_reference_names(dependency: &ManifestDependency) -> BTreeSet<String> {
    let mut names = BTreeSet::new();
    names.insert(normalize_dependency_reference(&dependency.identity));
    if let Some((_, name)) = dependency.identity.rsplit_once('.') {
        names.insert(normalize_dependency_reference(name));
    }
    if dependency.kind == ManifestDependencyKind::SourceControl {
        if let Some(name) = dependency
            .location
            .trim_end_matches('/')
            .trim_end_matches(".git")
            .rsplit('/')
            .next()
        {
            names.insert(normalize_dependency_reference(name));
        }
    }
    names
}

fn normalize_dependency_reference(name: &str) -> String {
    name.trim_end_matches(".git").to_ascii_lowercase()
}

pub(crate) fn parse_manifest_file_system_dependencies(
    manifest: &Value,
) -> Result<Vec<ManifestFileSystemDependency>> {
    let mut dependencies = Vec::new();
    let Some(items) = manifest.get("dependencies").and_then(Value::as_array) else {
        return Ok(dependencies);
    };

    for item in items {
        let Some(file_system) = item.get("fileSystem").and_then(Value::as_array) else {
            continue;
        };
        for dependency in file_system {
            let identity = dependency
                .get("identity")
                .and_then(Value::as_str)
                .ok_or_else(|| anyhow!("fileSystem dependency is missing identity"))?
                .to_string();
            let path = dependency
                .get("path")
                .and_then(Value::as_str)
                .ok_or_else(|| anyhow!("{identity} is missing path"))?
                .to_string();
            let name = dependency
                .get("nameForTargetDependencyResolutionOnly")
                .and_then(Value::as_str)
                .unwrap_or(&identity)
                .to_string();
            dependencies.push(ManifestFileSystemDependency {
                identity,
                name,
                path,
            });
        }
    }

    Ok(dependencies)
}

pub(crate) fn parse_requirement(requirement: &Value) -> Result<Requirement> {
    if let Some(exact) = requirement
        .get("exact")
        .and_then(Value::as_array)
        .and_then(|items| items.first())
        .and_then(Value::as_str)
    {
        return Ok(Requirement::Exact(Version::parse(exact)?));
    }
    if let Some(range) = requirement
        .get("range")
        .and_then(Value::as_array)
        .and_then(|items| items.first())
    {
        let lower = range
            .get("lowerBound")
            .and_then(Value::as_str)
            .ok_or_else(|| anyhow!("range is missing lowerBound"))?;
        let upper = range
            .get("upperBound")
            .and_then(Value::as_str)
            .ok_or_else(|| anyhow!("range is missing upperBound"))?;
        return Ok(Requirement::Range {
            lower: Version::parse(lower)?,
            upper: Version::parse(upper)?,
        });
    }
    if let Some(revision) = requirement
        .get("revision")
        .and_then(Value::as_array)
        .and_then(|items| items.first())
        .and_then(Value::as_str)
    {
        return Ok(Requirement::Revision(revision.to_string()));
    }
    if let Some(branch) = requirement
        .get("branch")
        .and_then(Value::as_array)
        .and_then(|items| items.first())
        .and_then(Value::as_str)
    {
        return Ok(Requirement::Branch(branch.to_string()));
    }
    bail!("unsupported requirement shape: {requirement}");
}

File: Sources/swifterpm/Manifest.swift
import Foundation

enum ManifestDependencyKind: Hashable, Sendable {
    case sourceControl
    case registry
}

struct ManifestDependency: Sendable {
    let identity: String
    let kind: ManifestDependencyKind
    let location: String
    let requirement: Requirement
}

struct ManifestFileSystemDependency: Sendable {
    let identity: String
    let name: String
    let path: String
}

enum Requirement: Sendable {
    case exact(SemVer)
    case range(lower: SemVer, upper: SemVer)
    case revision(String)
    case branch(String)
}

let manifestCacheFile = ".swifterpm-manifest.json"

func dumpPackage(packageDir: URL, disableSandbox: Bool) throws -> Any {
    let data = try dumpPackageJSON(packageDir: packageDir, disableSandbox: disableSandbox)
    return try JSONSerialization.jsonObject(with: data)
}

func dumpPackageJSON(packageDir: URL, disableSandbox: Bool) throws -> Data {
    if let cached = try readCachedManifest(packageDir: packageDir) {
        return cached
    }

    var args = ["package"]
    if disableSandbox {
        args.append("--disable-sandbox")
    }
    args.append("dump-package")
    let result = try runCommand("/usr/bin/swift", args, workingDirectory: packageDir)
    try? atomicWrite(result.stdout, to: packageDir.appendingPathComponent(manifestCacheFile))
    return result.stdout
}

private func readCachedManifest(packageDir: URL) throws -> Data? {
    let cache = packageDir.appendingPathComponent(manifestCacheFile)
    let manifest = packageDir.appendingPathComponent("Package.swift")
    guard FileManager.default.fileExists(atPath: cache.path) else { return nil }
    let cacheAttrs = try FileManager.default.attributesOfItem(atPath: cache.path)
    let manifestAttrs = try FileManager.default.attributesOfItem(atPath: manifest.path)
    guard let cacheDate = cacheAttrs[.modificationDate] as? Date,
          let manifestDate = manifestAttrs[.modificationDate] as? Date,
          cacheDate >= manifestDate
    else {
        return nil
    }
    return try Data(contentsOf: cache)
}

func parseManifestDependencies(_ manifest: Any) throws -> [ManifestDependency] {
    var dependencies: [ManifestDependency] = []
    guard let root = manifest as? [String: Any],
          let items = root["dependencies"] as? [[String: Any]]
    else {
        return dependencies
    }

    for item in items {
        if let sourceControl = item["sourceControl"] as? [[String: Any]] {
            for dependency in sourceControl {
                guard let identity = dependency["identity"] as? String else {
                    throw ToolError.message("sourceControl dependency is missing identity")
                }
                guard let location = parseSourceControlLocation(dependency) else {
                    throw ToolError.message("\(identity) is missing source-control location")
                }
                guard let requirementJSON = dependency["requirement"] else {
                    throw ToolError.message("\(identity) is missing requirement")
                }
                dependencies.append(ManifestDependency(
                    identity: identity,
                    kind: .sourceControl,
                    location: location,
                    requirement: try parseRequirement(requirementJSON)
                ))
            }
        }

        if let registry = item["registry"] as? [[String: Any]] {
            for dependency in registry {
                guard let identity = dependency["identity"] as? String else {
                    throw ToolError.message("registry dependency is missing identity")
                }
                guard let requirementJSON = dependency["requirement"] else {
                    throw ToolError.message("\(identity) is missing requirement")
                }
                dependencies.append(ManifestDependency(
                    identity: identity,
                    kind: .registry,
                    location: identity,
                    requirement: try parseRequirement(requirementJSON)
                ))
            }
        }
    }

    return dependencies
}

private func parseSourceControlLocation(_ dependency: [String: Any]) -> String? {
    guard let location = dependency["location"] as? [String: Any] else {
        return nil
    }
    if let remote = location["remote"] as? [[String: Any]],
       let first = remote.first,
       let url = first["urlString"] as? String
    {
        return url
    }
    if let local = location["local"] as? [String],
       let first = local.first
    {
        return first
    }
    return nil
}

func parseRequiredManifestDependencies(_ manifest: Any) throws -> [ManifestDependency] {
    let dependencies = try parseManifestDependencies(manifest)
    let references = activeDependencyReferences(manifest)
    if references.isEmpty {
        return []
    }
    return dependencies.filter { dependency in
        dependencyReferenceNames(dependency).contains { references.contains($0) }
    }
}

private func activeDependencyReferences(_ manifest: Any) -> Set<String> {
    guard let root = manifest as? [String: Any],
          let targets = root["targets"] as? [[String: Any]]
    else {
        return []
    }

    let targetNames = Set(targets.compactMap { $0["name"] as? String })
    var pendingTargets: [String] = []
    if let products = root["products"] as? [[String: Any]] {
        for product in products {
            if let targets = product["targets"] as? [String] {
                pendingTargets.append(contentsOf: targets)
            }
        }
    }
    for target in targets where target["type"] as? String == "test" {
        if let name = target["name"] as? String {
            pendingTargets.append(name)
        }
    }

    var references = Set<String>()
    var visitedTargets = Set<String>()

    while let targetName = pendingTargets.popLast() {
        guard visitedTargets.insert(targetName).inserted,
              let target = targets.first(where: { $0["name"] as? String == targetName }),
              let dependencies = target["dependencies"] as? [[String: Any]]
        else {
            continue
        }

        for dependency in dependencies {
            if let product = dependency["product"] as? [Any] {
                let productName = product.first as? String
                let packageName = product.count > 1 ? product[1] as? String : nil
                if let name = packageName ?? productName {
                    references.insert(normalizeDependencyReference(name))
                }
            }
            if let byName = dependency["byName"] as? [Any],
               let name = byName.first as? String
            {
                if targetNames.contains(name) {
                    pendingTargets.append(name)
                } else {
                    references.insert(normalizeDependencyReference(name))
                }
            }
        }
    }

    return references
}

private func dependencyReferenceNames(_ dependency: ManifestDependency) -> Set<String> {
    var names = Set<String>()
    names.insert(normalizeDependencyReference(dependency.identity))
    if let suffix = dependency.identity.split(separator: ".").last {
        names.insert(normalizeDependencyReference(String(suffix)))
    }
    if dependency.kind == .sourceControl,
       let name = dependency.location.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        .replacingOccurrences(of: ".git", with: "")
        .split(separator: "/")
        .last
    {
        names.insert(normalizeDependencyReference(String(name)))
    }
    return names
}

private func normalizeDependencyReference(_ name: String) -> String {
    let value = name.hasSuffix(".git") ? String(name.dropLast(4)) : name
    return value.lowercased()
}

func parseManifestFileSystemDependencies(_ manifest: Any) throws -> [ManifestFileSystemDependency] {
    guard let root = manifest as? [String: Any],
          let items = root["dependencies"] as? [[String: Any]]
    else {
        return []
    }

    var dependencies: [ManifestFileSystemDependency] = []
    for item in items {
        guard let fileSystem = item["fileSystem"] as? [[String: Any]] else { continue }
        for dependency in fileSystem {
            guard let identity = dependency["identity"] as? String else {
                throw ToolError.message("fileSystem dependency is missing identity")
            }
            guard let path = dependency["path"] as? String else {
                throw ToolError.message("\(identity) is missing path")
            }
            dependencies.append(ManifestFileSystemDependency(
                identity: identity,
                name: dependency["nameForTargetDependencyResolutionOnly"] as? String ?? identity,
                path: path
            ))
        }
    }
    return dependencies
}

func parseRequirement(_ requirement: Any) throws -> Requirement {
    guard let requirement = requirement as? [String: Any] else {
        throw ToolError.message("unsupported requirement shape: \(requirement)")
    }
    if let exact = requirement["exact"] as? [String], let value = exact.first {
        return .exact(try SemVer(value))
    }
    if let range = requirement["range"] as? [[String: Any]], let first = range.first {
        guard let lower = first["lowerBound"] as? String,
              let upper = first["upperBound"] as? String
        else {
            throw ToolError.message("range is missing lowerBound or upperBound")
        }
        return .range(lower: try SemVer(lower), upper: try SemVer(upper))
    }
    if let revision = requirement["revision"] as? [String], let value = revision.first {
        return .revision(value)
    }
    if let branch = requirement["branch"] as? [String], let value = branch.first {
        return .branch(value)
    }
    throw ToolError.message("unsupported requirement shape: \(requirement)")
}

func versionRange(for requirement: Requirement) -> VersionRange? {
    switch requirement {
    case let .exact(version):
        return .singleton(version)
    case let .range(lower, upper):
        return .between(lower, upper)
    case .revision, .branch:
        return nil
    }
}
